src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/Value.ReadTests.cs

```diff
@@ -442,5 +447,32 @@ private static void DeserializeLongJsonString(int stringLength)
             string str = JsonSerializer.Deserialize<string>(json);
             Assert.True(json.AsSpan(1, json.Length - 2).SequenceEqual(str.AsSpan()));
         }
+
+        [Theory]
+        [InlineData("23:59:59")]
+        [InlineData("23:59:59.9")]
+        [InlineData("23:59:59.9999999")]
+        [InlineData("9999999.23:59:59.9999999")]
+        [InlineData("-9999999.23:59:59.9999999")]
```

Comment:
Following Json.NET's example, the deserializer should probably also recognize `G/g` formatted timespans as well, e.g. `"9999999:23:59:59"`.

I would also include boundary values here, such as the serializations for `TimeSpan.MaxValue/MinValue`.

Reply:
It might be worth investigating if Json.NET supports other formats too.

Reply:
* I added some logic in `TimeSpanConverter` so that it will try 'g' format if 'c' is unsuccessful. People using 'c' will get optimal perf, people using 'g' will take a hit. Other option is to copy the `Utf8Parser.TryParse` code and tweak it for our needs. Those tweaks could lower the perf of 'c' case since it will need to also encompass the 'g' logic. Let me know what you think about this!

* It looks like Json.NET calls into `TimeSpan.ToString()` & `TimeSpan.Parse`:

   * https://github.com/JamesNK/Newtonsoft.Json/blob/f7e7bd05d9280f17993500085202ff4ea150564a/Src/Newtonsoft.Json/JsonTextWriter.cs#L751
 
   * https://github.com/JamesNK/Newtonsoft.Json/blob/f7e7bd05d9280f17993500085202ff4ea150564a/Src/Newtonsoft.Json/Utilities/ConvertUtils.cs#L253

   `TimeSpan.ToString()` uses the 'c' format so our write is compatible. `TimeSpan.Parse` is a little more fuzzy about what it accepts [which can lead to some surprises](https://blog.macrosssoftware.com/index.php/2020/06/04/timespan-parse240000-may-surprise-you/). I updated the tests to show that we round-trip with Json.NET for the most part, but break on that one known case. Let me know what you think about that!

Reply:
I would err towards using `TimeSpan.Parse` despite the issues, but first I'd be interested to see the performance impact versus the `Utf8Parser` fallback approach first for both 'c' and 'g' formats.

Reply:
I wrote a couple of quick benchmarks against NET5 here's how they look:

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19042.1052 (20H2/October2020Update)
Intel Core i9-9880H CPU 2.30GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK=5.0.301
  [Host]     : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT
  DefaultJob : .NET 5.0.7 (5.0.721.25508), X64 RyuJIT


|             Method |                Input |      Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------- |--------------------- |----------:|----------:|----------:|-------:|------:|------:|----------:|
|      TimeSpanParse | -1067(...)75808 [26] | 421.12 ns |  3.024 ns |  2.680 ns |      - |     - |     - |         - |
| Utf8ParserFallback | -1067(...)75808 [26] | 118.14 ns |  1.372 ns |  1.283 ns |      - |     - |     - |         - |
|      TimeSpanParse |             00:00:00 | 282.53 ns |  5.514 ns |  5.415 ns |      - |     - |     - |         - |
| Utf8ParserFallback |             00:00:00 |  82.21 ns |  0.682 ns |  0.569 ns |      - |     - |     - |         - |
|      TimeSpanParse |            1:2:00:00 | 620.75 ns | 12.390 ns | 16.540 ns | 0.0238 |     - |     - |     200 B |
| Utf8ParserFallback |            1:2:00:00 | 130.04 ns |  0.700 ns |  0.546 ns |      - |     - |     - |         - |

<details>
  <summary>Benchmark code</summary>

```csharp
using System;
using System.Buffers;
using System.Buffers.Text;
using System.Text;
using System.Text.Json;

using BenchmarkDotNet.Attributes;

namespace TimeSpanBenchmarks
{
	[MemoryDiagnoser]
	public class ParsingBenchmarks
	{
		private byte[]? _Utf8Data;

		[Params("00:00:00", "1:2:00:00", "-10675199.02:48:05.4775808")]
		public string? Input { get; set; }

		[GlobalSetup]
		public void GlobalSetup() => _Utf8Data = Encoding.UTF8.GetBytes($"\"{Input}\"");

		[Benchmark]
		public void TimeSpanParse()
		{
			Utf8JsonReader reader = new Utf8JsonReader(_Utf8Data);

			reader.Read();

			ReadOnlySpan<byte> value = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;

			int numberOfCharsRequired = value.Length;

			char[]? charArray = null;
			Span<char> valueAsChars;
			if (numberOfCharsRequired > 256)
			{
				charArray = ArrayPool<char>.Shared.Rent(numberOfCharsRequired);
				valueAsChars = charArray;
			}
			else
			{
				unsafe
				{
					char* ptr = stackalloc char[256];
					valueAsChars = new Span<char>(ptr, 256);
				}
			}

			int numberOfDecodedBytes = Encoding.UTF8.GetChars(value, valueAsChars);

			bool result = TimeSpan.TryParse(valueAsChars.Slice(0, numberOfDecodedBytes), out TimeSpan tsValue);

			if (charArray != null)
			{
				ArrayPool<char>.Shared.Return(charArray);
			}

			if (!result)
			{
				throw new JsonException();
			}
		}

		[Benchmark]
		public void Utf8ParserFallback()
		{
			Utf8JsonReader reader = new Utf8JsonReader(_Utf8Data);

			reader.Read();

			ReadOnlySpan<byte> source = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;

			bool result = Utf8Parser.TryParse(source, out TimeSpan tmpValue, out int bytesConsumed, 'c');

			/* Note: Utf8Parser.TryParse will return true for invalid input so
			   long as it starts with an integer. Example: "2021-06-18" or
			   "1$$$$$$$$$$". We need to check bytesConsumed to know if the
			   entire source was actually valid. */

			if (result && source.Length == bytesConsumed)
			{
				return;
			}

			result = Utf8Parser.TryParse(source, out tmpValue, out bytesConsumed, 'g');

			if (result && source.Length == bytesConsumed)
			{
				return;
			}

			throw new JsonException();
		}
	}
}
```
</details>

Reply:
Thanks, I think we can use Utf8Parser for now but could reconsider in the future if other use cases crop up. cc @layomia @steveharter for second opinions.

Reply:
Since there is a perf implication here, I think it's better to support just the 'c' format for now since it is the default output of `Newtonsoft.Json` and should cover most scenarios. We can adjust based on user feedback. This is in keeping with support for other primitives where we use just one format. It is easy to write a custom converter for other formats.

Reply:
Removed the 'g' fallback